ssl/ssl_lib.c

```diff
@@ -4074,59 +4070,168 @@ int SSL_CTX_set_ct_validation_callback(SSL_CTX *ctx, ct_validation_cb callback,
             TLSEXT_TYPE_signed_certificate_timestamp)) {
         SSLerr(SSL_F_SSL_CTX_SET_CT_VALIDATION_CALLBACK,
                SSL_R_CUSTOM_EXT_HANDLER_ALREADY_INSTALLED);
-        goto err;
+        return 0;
     }
 
     ctx->ct_validation_callback = callback;
     ctx->ct_validation_callback_arg = arg;
-    ret = 1;
-err:
-    return ret;
+    return 1;
+}
+
+int SSL_ct_is_enabled(const SSL *s)
+{
+    return s->ct_validation_callback != NULL;
+}
+
+int SSL_CTX_ct_is_enabled(const SSL_CTX *ctx)
+{
+    return ctx->ct_validation_callback != NULL;
+}
+
+static void *ssl_ct_arg(int vtype)
+{
+    static int permissive = SSL_CT_VALIDATION_PERMISSIVE;
+    static int strict = SSL_CT_VALIDATION_STRICT;
+
+    switch (vtype) {
+    default:
+        SSLerr(SSL_F_SSL_CT_ARG, SSL_R_INVALID_CT_VALIDATION_TYPE);
+        return NULL;
+    case SSL_CT_VALIDATION_PERMISSIVE:
+        return &permissive;
+    case SSL_CT_VALIDATION_STRICT:
+        return &strict;
+    }
+}
+
+static int ssl_ct_cb(const CT_POLICY_EVAL_CTX *ctx,
+                     const STACK_OF(SCT) *scts, void *arg)
+{
+    int count;
+    int i;
+
+    switch (arg == NULL ? -1 : *(int *)arg) {
```

Comment:
The constants are not and should not be negative. Since Rob is considering replacing this with function pointers, I'll leave this to the follow on PR.